metabase/metabase
# Skip long values in JSON unfolding query

Resolves https://github.com/metabase/metabase/issues/45163

### Why OOM errors can happen during JSON unfolding

In `describe-json-fields`, we fetch a sample of 500 rows of data into a ResultSet object from the database. We select all the JSON columns in the table at once for these rows. If there are sufficiently large JSON values among this sample we exhaust the heap space. This is all without even starting to parse the values. We could decrease the fetch size and fetch the data in batches, but doing that will make sync slower.

### The solution

This PR limits the size of each value that can be returned from the query.

Until recently (after https://github.com/metabase/metabase/pull/42638) we had a limit of 50,000 characters in a JSON string before we would attempt to parse a JSON value. This PR restores that 50,000 character limit, but in the query instead. If there is a JSON value that is longer than 50,000 characters, the query will return NULL in its place.

This PR technically constitutes a breaking change because it may result in some fields becoming inactive that would previously would be active. I still think we should backport it to 49 and 50, because that scenario seems relatively unlikely, the behaviour is the same as before https://github.com/metabase/metabase/pull/42638, and it is outweighed by the benefit of fixing the OOM errors.

### How to verify manually

```
;; launch Metabase with 4 GB of maximum heap space by setting -Xmx4g option in deps.edn

(require
 '[cheshire.core :as json]
 '[clojure.java.jdbc :as jdbc]
 '[metabase.driver.sql-jdbc.connection :as sql-jdbc.conn]
 '[metabase.sync :as sync]
 '[metabase.test :as mt]
 '[toucan2.core :as t2])

;; Create a new test db:
(mt/defdataset test-db [])
(mt/with-driver :postgres (mt/dataset test-db (def db (mt/db))))

;; Set up the table:
(let [j (json/encode {"a" (repeat 10000000 "x")})]
  (with-open [conn (jdbc/get-connection (sql-jdbc.conn/db->pooled-connection-spec db))]
    (jdbc/execute! {:connection conn} "CREATE TABLE IF NOT EXISTS large_jsonb_test (id SERIAL PRIMARY KEY, data JSONB)")
    (jdbc/execute! {:connection conn} "TRUNCATE TABLE large_jsonb_test")
    (dotimes [_ 100]
      (jdbc/execute! {:connection conn} [(str "INSERT INTO large_jsonb_test ( data ) VALUES (?::jsonb)") j]))))

(-> (:bytes (first (t2/query db "SELECT sum(pg_column_size(data)) as bytes FROM large_jsonb_test;")))
    (/ 1024 1024)
    float)
;; => 140 MB or so

;; the sync-fields step will fail with an OOM error before the change, and succeed after the change.
(sync/sync-database! db {:scan :schema})
```

This should be backported to 49. 
The new Druid JDBC driver will be implemented separately (issue: https://github.com/metabase/metabase/issues/46137), because it's currently difficult to use a new test dataset.
